Assets/Scripts/MovePlayer.cs

```diff
@@ -0,0 +1,213 @@
+    }
+
+    // Update is called once per frame
+    void Update()
+    {
+        GetInput();
+    }
+
+    private void FixedUpdate()
+    {
+        if (canMove && shouldMove && !moving)
+            StartCoroutine(MovementRoutine());
+    }
+
+    private void OnTriggerEnter2D(Collider2D collision)
+    {
+        if (collision.gameObject.tag == slowTerrainTag)
+            slowMove = !slowMove;
+    }
+
+    /// <summary>
+    /// Gets player input
+    /// </summary>
+    private void GetInput()
+    {
+        if (Input.GetAxis("Vertical") < 0)
+        {
+            shouldMove = true;
+            currentDirection = directions.Down;
+        }
+        else if (Input.GetAxis("Vertical") > 0)
+        {
+            shouldMove = true;
+            currentDirection = directions.Up;
+        }
+        else if (Input.GetAxis("Horizontal") < 0)
+        {
+            shouldMove = true;
+            currentDirection = directions.Left;
+        }
+        else if (Input.GetAxis("Horizontal") > 0)
+        {
+            shouldMove = true;
+            currentDirection = directions.Right;
+        }
+        else
+            shouldMove = false;
+    }
+
+    /// <summary>
+    /// Adjusts the player to fit exactly into a tile
+    /// </summary>
+    private void SnapToGrid()
+    {
+        int positionX = Mathf.RoundToInt(transform.position.x);
+        int positionY = Mathf.RoundToInt(transform.position.y);
+        transform.position = new Vector3(positionX, positionY);
+    }
+
+
+    private void PlayWalkAnim(directions movementDirection, bool slowTerrain)
+    {
+        playerAnimator.SetBool("Slow", slowTerrain);
+        switch (movementDirection)
+        {
+            case directions.Down:
+                playerAnimator.SetTrigger("Move Down");
+                break;
+            case directions.Up:
+                playerAnimator.SetTrigger("Move Up");
+                break;
+            case directions.Left:
+                playerAnimator.SetTrigger("Move Left");
+                break;
+            case directions.Right:
+                playerAnimator.SetTrigger("Move Right");
+                break;
+        }
+    }
+
+    /// <summary>
+    /// Raycasts for collision in the current direction before moving
+    /// </summary>
+    /// <returns>True if there is something in the way, else false.</returns>
+    private bool CheckCollisionRaycast(directions attemptedMoveDirection)
+    {
+        RaycastHit2D hit = Physics2D.Raycast(collisionRaycastPoint.position, movementVectors[(int)attemptedMoveDirection],1f, collisionLayerMask);
+        return hit.collider != null;
+    }
+
+    /// <summary>
+    /// Raycasts for slow terrain in the current direction before moving
+    /// </summary>
+    /// <returns>True if there is slow terrain in the way, else false.</returns>
+    private bool CheckSlowTerrainRaycast(directions attemptedMoveDirection)
```

Comment:
What's the intended design:

- Player moves slow if moving into slow terrain?

- Player moves slow if moving out of slow terrain?

Reply:
Once I know the design intent I can think about what implementation makes sense

Reply:
The intent is that the player moves slowly if moving into slow terrain, but not out of it.